pulsar/internal/connection.go

```diff
@@ -656,18 +662,20 @@ func (c *connection) Close() {
 
 	c.log.Info("Connection closed")
 	c.state = connectionClosed
-	if c.cnx != nil {
-		c.cnx.Close()
-	}
+	c.TriggerClose()
 	c.pingTicker.Stop()
 	c.pingCheckTicker.Stop()
 
 	for _, listener := range c.listeners {
 		listener.ConnectionClosed()
 	}
 
-	for _, req := range c.pendingReqs {
+	if c.runLoopStoppedCh != nil {
```

Comment:
can you reuse the close channel here? Why create another channel?

Reply:
Do you mean reusing the `closeCh?

It seems that the `closeCh` is a signal to instruct the `run()` loop to exit. However we would not know when the `run()` has finished its job without a dedicated signal.

Therefore, the `runLoopStoppedCh` is the dedicated signal and is responsible to indicating the `run()` loop stopped.
When `runLoopStoppedCh` closed, we are able do further cleanup resources shared with the `run()` loop.

Reply:
Why does the run loop have to exit before removing the pending callback? How could a callback be called twice?

Reply:
As @wolfstudy pointed out, the problem `internalWriteData` will call on `Close()`
https://github.com/apache/pulsar-client-go/blob/master/pulsar/internal/connection.go#L331

Then the run loop continue, and the callbacks in the `pendingReqs` have already been called. At this point, the run loop can go into `incomingCmdCh` branch even if we have closed the `c.closeCh` then the callbacks in the `pendingReqs` can be called second time.

Reply:
So the issue is connection.Close() is called multiple times? 

After Close() is the connection struct ever used again? 

If we want to ensure Close() is only called once can we use a sync.Once and do something similar to what the consumers do?

https://github.com/apache/pulsar-client-go/blob/master/pulsar/consumer_impl.go#L362

Reply:
Not exactly, the issue is the run loop can go into the `<-incomingCmdCh` branch and then call a callback in the `pendingReqs` which has already been called by the `Close()`. The `Close()` could come from previous iterations of the run loop or from another goroutine, therefore, wrapping a `sync.Once` on it can't stop it racing on the `pendingReqs` with the run loop.

However we do need to wrap `sync.Once` on the `TriggerClose()` since it is a public method and the `select` check before `close(c.closeCh)` is not thread safe as well.

Sorry for the ambiguous explanation about the run loop. I write a code snippet to simulate the original issue #239:

```golang
package internal

import (
	"net"
	"runtime/debug"
	"sync"
	"testing"
	"time"

	"github.com/apache/pulsar-client-go/pulsar/internal/pb"
	log "github.com/sirupsen/logrus"
)

func TestConnection_Close(t *testing.T) {
	defer func() {
		if r := recover(); r != nil {
			t.Fatalf("%v %v", r, string(debug.Stack()))
		}
	}()
	for i := 0; i < 10000; i++ {
		pi, po := net.Pipe()
		cnx := fakeConn(po)

		// fake pending cmd from the server
		req, cmd := fakeReqCmd()
		cnx.pendingReqs[*req.id] = req
		cnx.incomingCmdCh <- &incomingCmd{cmd: cmd}

		// make internalWriteData fail
		pi.Close()
		cnx.writeRequestsCh <- []byte{' '}

		cnx.run()
	}
}

func fakeConn(conn net.Conn) *connection {
	cnx := &connection{
		cnx:                  conn,
		log:                  log.WithField("remote_addr", "test"),
		lastDataReceivedTime: time.Now(),
		pingTicker:           time.NewTicker(keepAliveInterval),
		pingCheckTicker:      time.NewTicker(keepAliveInterval),
		pendingReqs:          make(map[uint64]*request),
		closeCh:              make(chan interface{}),
		incomingCmdCh:        make(chan *incomingCmd, 10),
		writeRequestsCh:      make(chan []byte, 10),
	}
	cnx.reader = newConnectionReader(cnx)
	cnx.cond = sync.NewCond(cnx)
	return cnx
}

func fakeReqCmd() (*request, *pb.BaseCommand) {
	id, t := uint64(1), pb.BaseCommand_SUCCESS
	ch := make(chan interface{})

	return &request{
		id: &id,
		callback: func(command *pb.BaseCommand, err error) {
			close(ch)
		},
	}, &pb.BaseCommand{Type: &t, Success: &pb.CommandSuccess{RequestId: &id}}
}
```

and the result:

```
=== RUN   TestConnection_Close
time="2020-05-16T13:09:16+08:00" level=warning msg="Failed to write on connection" error="io: read/write on closed pipe" remote_addr=test
time="2020-05-16T13:09:16+08:00" level=info msg="Connection closed" remote_addr=test
time="2020-05-16T13:09:16+08:00" level=info msg="Error reading from connection" error="Short read when reading frame size" remote_addr=test
time="2020-05-16T13:09:16+08:00" level=warning msg="Failed to write on connection" error="io: read/write on closed pipe" remote_addr=test
time="2020-05-16T13:09:16+08:00" level=info msg="Connection closed" remote_addr=test
    TestConnection_Close: connection_test.go:17: close of closed channel goroutine 6 [running]:
        runtime/debug.Stack(0xc00016fab0, 0x145b480, 0x158fda0)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
        github.com/apache/pulsar-client-go/pulsar/internal.TestConnection_Close.func1(0xc0001350e0)
        	/Users/ruian/Code/go/pulsar-client-go/pulsar/internal/connection_test.go:17 +0x57
        panic(0x145b480, 0x158fda0)
        	/usr/local/go/src/runtime/panic.go:967 +0x15d
        github.com/apache/pulsar-client-go/pulsar/internal.fakeReqCmd.func1(0xc000244380, 0x0, 0x0)
        	/Users/ruian/Code/go/pulsar-client-go/pulsar/internal/connection_test.go:61 +0x2a
        github.com/apache/pulsar-client-go/pulsar/internal.(*connection).handleResponse(0xc00021a900, 0x1, 0xc000244380)
        	/Users/ruian/Code/go/pulsar-client-go/pulsar/internal/connection.go:493 +0xb3
        github.com/apache/pulsar-client-go/pulsar/internal.(*connection).internalReceivedCommand(0xc00021a900, 0xc000244380, 0x0, 0x0)
        	/Users/ruian/Code/go/pulsar-client-go/pulsar/internal/connection.go:400 +0x210
        github.com/apache/pulsar-client-go/pulsar/internal.(*connection).run(0xc00021a900)
        	/Users/ruian/Code/go/pulsar-client-go/pulsar/internal/connection.go:325 +0x304
        github.com/apache/pulsar-client-go/pulsar/internal.TestConnection_Close(0xc0001350e0)
        	/Users/ruian/Code/go/pulsar-client-go/pulsar/internal/connection_test.go:33 +0xf8
        testing.tRunner(0xc0001350e0, 0x14fd630)
        	/usr/local/go/src/testing/testing.go:992 +0xdc
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1043 +0x357
--- FAIL: TestConnection_Close (0.00s)
FAIL

Process finished with exit code 1

```

Reply:
I'm concerned with channel `runLoopStoppedCh` because I think it complicates the code (there should at least be some comments). Is there another way to solve this?

If I understand correctly callbacks from this `c.pendingReqs` could be called multiple times due to the go routine used by the run() loop and another go routine that might have called Close()?

Can we use move this code into the run loop? 
```
for id, req := range c.pendingReqs {
	req.callback(nil, errors.New("connection closed"))
	delete(c.pendingReqs, id)
}
```
That go routine is the only thing that accesses the `pendingReqs` map (excluding the Close()). If we moved the above code from Close() into the run() we don't need a stopChannel and I think it simplifies the code and fixes any race conditions with callbacks.

Reply:
Yes, I agree that moving the above code into the run loop is better and simpler. Actually it was my first attempt to solve this problem. 

However, I am not sure we can move only the above code from `Close()` into the run loop. There are still some cleanup on the `c.listeners` and `c.consumerHandlers` maps in the following code. Should we move them into the run loop together to prevent creating another race condition?

If we move one of them into the run loop, the `Close()` will become an async operation which I am not sure whether it will also break other code.

Any feedback?

Reply:
I'm not sure if order is needed (pendingReqs close callbacks must happen before the others c.listeners and c.consumerHandlers callbacks) maybe @merlimat would know better. 

I think moving just the pendingReqs into the run loops is needed. The other structures are protected with locks. Have you tried moving the pendingReqs code from Close() to the run loop and then running some tests?

Reply:
Yes, moving just the pendingReqs into the run loops will still pass the existing tests, but it is really hard to write new race tests without to be flaky.

I would suggest we keep cleanup code at only one place, and they would be better triggered by the run loop. However the `Close()` would become an async operation.

Hi @merlimat, would you have any feedback?

Reply:
After tracing the call path of the `Close()`, I found that it is called directly by the `Close()` in the `client_impl.go`. 

When performing a graceful shutdown, although normally we will first call the `producer.Close()/consumer.Close()` and then call `client.Close()` at the end. I think it is still a good idea to keep the `Close()` as a sync operation to remain similar behavior between the `producer.Close()/consumer.Close()`.

We could move the cleanup code of `c.pendingReqs`, `c.listeners` and `c.consumerHandlers` into the run loop. However in that case, I would suggest keeping the `runLoopStoppedCh` channel to inform the `Close()` that the cleanup process has finished. 

What do you think? @cckellogg

Reply:
Do you see the run loop ever getting stuck for a period of time when close is called? I think we should probably only pendingReqs cleanup in the run loop since that map is only accessed in that go routine . We could keep the others where they are since they are protected by locks. We just need to add some comments on why this is the case and why the runLoopStoppedCh is needed.

Thoughts?

Reply:
> Do you see the run loop ever getting stuck for a period of time when close is called?

No, they are separated originally. To keep the original behavior, the runLoopStoppedCh is not needed if we move the pendingReqs cleanup at the end of run loop.

I just updated the PR which moves pendingReqs cleanup into the run loop and adds some notices.

Please let me know if it looks good to you.